crates/hyperdrive-math/src/lp.rs

```diff
@@ -600,6 +621,540 @@ impl State {
+
+        // Calculate the max bond amount. if the calculation fails, we return a
+        // failure flag. if the calculation succeeds but the max bond amount
+        // is zero, then we return a failure flag since we can't divide by zero.
+        let max_bond_amount = self.calculate_max_buy_bonds_out_safe()?;
+        if max_bond_amount == fixed!(0) {
+            return Err(eyre!("Max bond amount zero."));
+        }
+
+        // We can solve for the maximum share reserves delta in one shot using
+        // the fact that the maximum amount of bonds that can be purchased is
+        // linear with respect to the scaling factor applied to the reserves.
+        // In other words, if s > 0 is a factor scaling the reserves, we have
+        // the following relationship:
+        //
+        // y_out^max(s * z, s * y, s * zeta) = s * y_out^max(z, y, zeta)
+        //
+        // We solve for the maximum share reserves delta by finding the scaling
+        // factor that results in the maximum amount of bonds that can be
+        // purchased being equal to the net curve trade. We can derive this
+        // maximum using the linearity property mentioned above as follows:
+        //
+        // y_out^max(s * z, s * y, s * zeta) - netCurveTrade = 0
+        //                        =>
+        // s * y_out^max(z, y, zeta) - netCurveTrade = 0
+        //                        =>
+        // s = netCurveTrade / y_out^max(z, y, zeta)
+        let max_scaling_factor = net_curve_trade.div_up(max_bond_amount);
+
+        // Using the maximum scaling factor, we can calculate the maximum share
+        // reserves delta as:
+        //
+        // maxShareReservesDelta = z * (1 - s)
+        if max_scaling_factor <= fixed!(1e18) {
+            let max_share_reserves_delta =
+                (fixed!(1e18) - max_scaling_factor).mul_down(self.share_reserves());
+
+            // If the maximum share reserves delta is greater than the idle, then
+            // the maximum share reserves delta is equal to the idle.
+            if max_share_reserves_delta > idle {
+                return Ok(idle);
+            }
+            return Ok(max_share_reserves_delta);
+        } else {
+            return Err(eyre!("max_scaling_factor greater than 1."));
+        }
+    }
+
+    /// Calculates the amount of withdrawal shares that can be redeemed
+    /// given an amount of shares to remove from the share reserves.
+    /// Assuming that dz is the amount of shares to remove from the
+    /// reserves and dl is the amount of LP shares to be burned, we can
+    /// derive the calculation as follows:
+    ///
+    ///  PV(0) / l = PV(dx) / (l - dl)
+    ///                =>
+    ///  dl = l - l * (PV(dx) / PV(0))
+    ///
+    ///  We round this calculation up to err on the side of slightly too
+    ///  many withdrawal shares being redeemed.
+    fn calculate_distribute_express_idle_withdrawal_shares_redeemed(
```

Comment:
is that supposed to be "express_idle" and not "excess_idle"?

Reply:
express idle sounds better 😂